File: backend/app/services/merchant_settings_service.py
from sqlalchemy import select
from sqlalchemy.exc import IntegrityError, SQLAlchemyError
from sqlalchemy.orm import Session

from backend.app.models.merchant_settings_model import MerchantSettings
from backend.app.schemas.merchant_settings_schema import (
    AutomationSettingsUpdate,
    CategoryRulesUpdate,
    EvidenceSettingsUpdate,
    MerchantProfileUpdate,
    MerchantSettingsResponse,
    MerchantSettingsUpdate,
    NotificationSettingsUpdate,
    RiskSettingsUpdate,
)


DEFAULT_SETTINGS = {
    "business_name": "ReturnIQ Merchant",
    "support_email": "",
    "support_phone": "",
    "website_url": "",
    "timezone": "Asia/Kolkata",
    "currency": "INR",
    "low_risk_max": 29,
    "medium_risk_max": 59,
    "high_risk_max": 79,
    "human_review_threshold": 60,
    "auto_approval_enabled": True,
    "auto_approval_max_score": 25,
    "auto_approval_max_amount": 2500.0,
    "auto_rejection_enabled": False,
    "auto_rejection_min_score": 90,
    "require_evidence": True,
    "evidence_minimum_images": 1,
    "evidence_required_above_amount": 3000.0,
    "allow_jpeg": True,
    "allow_png": True,
    "allow_webp": True,
    "maximum_upload_size_mb": 10,
    "email_notifications": True,
    "high_risk_alerts": True,
    "review_assignment_alerts": True,
    "daily_summary_enabled": True,
    "weekly_report_enabled": False,
    "notification_email": "",
    "default_return_window_days": 30,
    "returnless_refund_enabled": False,
    "returnless_refund_max_amount": 500.0,
    "manual_override_enabled": True,
    "require_override_remarks": True,
    "product_category_rules": {},
}


def _get_settings_record(
    db: Session,
    owner_id: str,
) -> MerchantSettings | None:
    statement = select(MerchantSettings).where(
        MerchantSettings.owner_id == owner_id
    )
    return db.execute(statement).scalar_one_or_none()


def _build_default_values(email: str | None = None) -> dict:
    values = DEFAULT_SETTINGS.copy()
    values["product_category_rules"] = {}

    if email:
        values["support_email"] = email
        values["notification_email"] = email

    return values


def get_or_create_merchant_settings(
    db: Session,
    owner_id: str,
    owner_email: str | None = None,
) -> MerchantSettings:
    existing_settings = _get_settings_record(db, owner_id)

    if existing_settings is not None:
        return existing_settings

    settings = MerchantSettings(
        owner_id=owner_id,
        **_build_default_values(owner_email),
    )

    try:
        db.add(settings)
        db.commit()
        db.refresh(settings)
        return settings

    except IntegrityError:
        db.rollback()
        existing_settings = _get_settings_record(db, owner_id)
        if existing_settings is not None:
            return existing_settings
        raise

    except SQLAlchemyError:
        db.rollback()
        raise


def _update_settings_fields(
    db: Session,
    settings: MerchantSettings,
    values: dict,
) -> MerchantSettings:
    for field_name, field_value in values.items():
        if hasattr(settings, field_name):
            setattr(settings, field_name, field_value)

    try:
        db.add(settings)
        db.commit()
        db.refresh(settings)
        return settings

    except SQLAlchemyError:
        db.rollback()
        raise


def _update(
    db: Session,
    owner_id: str,
    owner_email: str | None,
    payload,
) -> MerchantSettings:
    settings = get_or_create_merchant_settings(
        db,
        owner_id,
        owner_email,
    )
    return _update_settings_fields(
        db,
        settings,
        payload.model_dump(),
    )


def update_all_settings(db, owner_id, owner_email, payload: MerchantSettingsUpdate):
    return _update(db, owner_id, owner_email, payload)


def update_merchant_profile(db, owner_id, owner_email, payload: MerchantProfileUpdate):
    return _update(db, owner_id, owner_email, payload)


def update_risk_settings(db, owner_id, owner_email, payload: RiskSettingsUpdate):
    return _update(db, owner_id, owner_email, payload)


def update_automation_settings(
    db,
    owner_id,
    owner_email,
    payload: AutomationSettingsUpdate,
):
    return _update(db, owner_id, owner_email, payload)


def update_evidence_settings(
    db,
    owner_id,
    owner_email,
    payload: EvidenceSettingsUpdate,
):
    return _update(db, owner_id, owner_email, payload)


def update_notification_settings(
    db,
    owner_id,
    owner_email,
    payload: NotificationSettingsUpdate,
):
    return _update(db, owner_id, owner_email, payload)


def update_category_rules(
    db,
    owner_id,
    owner_email,
    payload: CategoryRulesUpdate,
):
    return _update(db, owner_id, owner_email, payload)


def reset_merchant_settings(
    db: Session,
    owner_id: str,
    owner_email: str | None,
) -> MerchantSettings:
    settings = get_or_create_merchant_settings(
        db,
        owner_id,
        owner_email,
    )
    return _update_settings_fields(
        db,
        settings,
        _build_default_values(owner_email),
    )


def serialize_merchant_settings(
    settings: MerchantSettings,
) -> MerchantSettingsResponse:
    return MerchantSettingsResponse.model_validate(settings)

File: frontend/src/services/merchantSettingsService.ts
import api from "./api"

export interface MerchantSettings {
  id: string
  owner_id: string

  business_name: string
  support_email: string
  support_phone: string
  website_url: string
  timezone: string
  currency: string

  low_risk_max: number
  medium_risk_max: number
  high_risk_max: number
  human_review_threshold: number

  auto_approval_enabled: boolean
  auto_approval_max_score: number
  auto_approval_max_amount: number
  auto_rejection_enabled: boolean
  auto_rejection_min_score: number

  require_evidence: boolean
  evidence_minimum_images: number
  evidence_required_above_amount: number
  allow_jpeg: boolean
  allow_png: boolean
  allow_webp: boolean
  maximum_upload_size_mb: number

  email_notifications: boolean
  high_risk_alerts: boolean
  review_assignment_alerts: boolean
  daily_summary_enabled: boolean
  weekly_report_enabled: boolean
  notification_email: string

  default_return_window_days: number
  returnless_refund_enabled: boolean
  returnless_refund_max_amount: number
  manual_override_enabled: boolean
  require_override_remarks: boolean

  product_category_rules: Record<string, unknown>

  created_at: string
  updated_at: string
}

export type MerchantProfilePayload = Pick<
  MerchantSettings,
  | "business_name"
  | "support_email"
  | "support_phone"
  | "website_url"
  | "timezone"
  | "currency"
>

export type RiskSettingsPayload = Pick<
  MerchantSettings,
  | "low_risk_max"
  | "medium_risk_max"
  | "high_risk_max"
  | "human_review_threshold"
>

export type AutomationSettingsPayload = Pick<
  MerchantSettings,
  | "auto_approval_enabled"
  | "auto_approval_max_score"
  | "auto_approval_max_amount"
  | "auto_rejection_enabled"
  | "auto_rejection_min_score"
  | "returnless_refund_enabled"
  | "returnless_refund_max_amount"
  | "default_return_window_days"
  | "manual_override_enabled"
  | "require_override_remarks"
>

export type EvidenceSettingsPayload = Pick<
  MerchantSettings,
  | "require_evidence"
  | "evidence_minimum_images"
  | "evidence_required_above_amount"
  | "allow_jpeg"
  | "allow_png"
  | "allow_webp"
  | "maximum_upload_size_mb"
>

export type NotificationSettingsPayload = Pick<
  MerchantSettings,
  | "email_notifications"
  | "high_risk_alerts"
  | "review_assignment_alerts"
  | "daily_summary_enabled"
  | "weekly_report_enabled"
  | "notification_email"
>

export interface ResetSettingsResponse {
  message: string
  settings: MerchantSettings
}

export const getMerchantSettings =
  async (): Promise<MerchantSettings> => {
    const response =
      await api.get<MerchantSettings>("/settings")

    return response.data
  }

export const updateMerchantProfile = async (
  payload: MerchantProfilePayload,
): Promise<MerchantSettings> => {
  const response = await api.patch<MerchantSettings>(
    "/settings/profile",
    payload,
  )

  return response.data
}

export const updateRiskSettings = async (
  payload: RiskSettingsPayload,
): Promise<MerchantSettings> => {
  const response = await api.patch<MerchantSettings>(
    "/settings/risk",
    payload,
  )

  return response.data
}

export const updateAutomationSettings = async (
  payload: AutomationSettingsPayload,
): Promise<MerchantSettings> => {
  const response = await api.patch<MerchantSettings>(
    "/settings/automation",
    payload,
  )

  return response.data
}

export const updateEvidenceSettings = async (
  payload: EvidenceSettingsPayload,
): Promise<MerchantSettings> => {
  const response = await api.patch<MerchantSettings>(
    "/settings/evidence",
    payload,
  )

  return response.data
}

export const updateNotificationSettings = async (
  payload: NotificationSettingsPayload,
): Promise<MerchantSettings> => {
  const response = await api.patch<MerchantSettings>(
    "/settings/notifications",
    payload,
  )

  return response.data
}

export const resetMerchantSettings =
  async (): Promise<ResetSettingsResponse> => {
    const response =
      await api.post<ResetSettingsResponse>(
        "/settings/reset",
      )

    return response.data
  }
